[e2e tests] Fix flaky create grouped product test

veljkho opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #50820

These tests were pretty non flaky but sometimes flaky. As Buildkite says, over 98% reliability score.

The flaky part was waiting for a response. I added a regression step instead of waiting for a response to fix the flakiness and increase coverage. It worked well locally, both with and without Gutenberg—100% success.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

```
pnpm test:e2e tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js --repeat-each=10
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:
```diff
@@ -72,11 +72,9 @@ test.describe( 'General tab', { tag: '@gutenberg' }, () => {
 				.getByText( 'Grouped product' )
 				.click();
 
-			await page.waitForResponse(
-				( response ) =>
-					response.url().includes( '/wp-json/wc/v3/products/' ) &&
-					response.status() === 200
-			);
+			await expect(
+				page.getByLabel( 'Dismiss this notice' )
+			).toContainText( 'Product type changed.' );
```
When we switch the product type, in this case to a grouped type, there's a saving in progress (draft). I added waiting for the notice of successful saving and then proceeding further.

adimoldovan replied:
I don't agree that this change will fix the flakiness. If the `/wp-json/wc/v3/products/` request fails the notice will not be displayed either. We need to understand why the the `/wp-json/wc/v3/products/`failed first.

veljkho replied:
The notice was indeed present even after failing to verify the `waitForResponse`. I believe this might be a rare race condition, as it happens very infrequently, and the test’s reliability score remains quite high.

To address this, I’ve moved the `waitForResponse` earlier in the process and adjusted the code accordingly. Could you please review the latest changes and let me know your thoughts?

juliaamosova replied:
>To address this, I’ve moved the `waitForResponse` earlier in the process and adjusted the code accordingly.

@veljkho we still haven't figured out why waitForResponse may be failing in its original place (after the user clicks on _Change product type_ and selects _Grouped product_). As @adimoldovan mentioned above, this is a key to understanding why the test is flaky. 

What are the reasons you moved `waitForResponse` after the user clicks on _General_ tab? 

I inspected API calls being made when the user selects _Grouped product_ and I indeed saw the POST request to create the product:

<img width="1508" alt="api_response" src="https://github.com/user-attachments/assets/e255d25b-7e75-4923-adfb-db825dc71acd">

With that said, we need to keep checking for a successful response in its original place to ensure the product draft gets created before proceeding. 

I further researched how to use `waitForResponse` properly and I came across [this article](https://www.workwithloop.com/blog/our-1-solution-to-playwright-flakiness-waitforresponse-waitforrequest-promises) that suggests using [Promise ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)with `waitForResponse`. If we use `waitForResponse` without a promise - quoting from the article further:
>There is a risk that the API may have already triggered and received the response by the time Playwright starts waiting for it. This can occur if the API is particularly fast, or if there is a delay in the execution of the test script due to other factors. If this happens, the waitForResponse will time out and the test will fail.

I believe this is what happens in our case - judging by the [video of the failed test you shared here](https://github.com/woocommerce/woocommerce/pull/53310#issuecomment-2510936439) and then [mentioned](https://github.com/woocommerce/woocommerce/pull/53310#issuecomment-2511723070) that it was a case when the test actually failed despite the video showing that the flow was successful. 

As a next step here, I would try to use a Promise to make the check for API response to mitigate the risk of the test failing due to flakiness in this place. 

I also see that you added the check for the notice `Product type changed` to appear after the change of product type. As we saw in the video you shared, the notice appeared even after the `waitForResponse` failed so the check for the notice doesn't help here and can be removed.

veljkho replied:
Thanks @juliaamosova for this constructive feedback that I was looking to receive all the time. I really appreciate the time you put into this.

> this is a key to understanding why the test is flaky.

In my investigation, I already understood that this was due to a race condition and the inappropriate use of `waitForResponse`. Even with the original approach, we are dealing with a very low flakiness report according to BK, where we have currently only 1 instance out of 100 last runs and the retry succeeded in that instance so we are already spending too much time on this, but I feel we can make it flawless so that's why we are moving forward.

> What are the reasons you moved waitForResponse after the user clicks on General tab?

This is exactly how to fix this. I think that the API call was successful, but the `waitForResponse` failed to catch it on time and produced a race condition issue.

I have observed it locally to prove this. [Here is the video with successful API call and failing](https://d.pr/i/7SQXVm) `waitForResponse`. [This is the code added](https://d.pr/i/VJiD0X). With this, we have successfully received API but it was called later in the process so it failed to assert it. If I remove that waitForResponse, the test will be successful showing that my UI based assertion helped to solve the flakiness.

My intention is to illustrate why my change to wait for the notice offers an improvement over the original use of `waitForResponse`. However, **I strongly value collaborative thinking and teamwork**, so I don’t want to focus solely on my approach while excluding others’ contributions. There’s always a way to integrate both perspectives, and I am actively working on achieving that. I would leave notice assertion as it stabilizes the test and verifies important UI change while switching product type.

> As a next step here, I would try to use a Promise to make the check for API response to mitigate the risk of the test failing due to flakiness in this place.

I’m open to exploring other approaches, as there are likely several ways to address this. That said, I’d appreciate it if you could review and test my code as well. I’ve tested it extensively, running it a [hundred times](https://d.pr/i/uDAMmm) with great success.
I’m also keen to hear your thoughts and have you evaluate my findings to ensure we’re aligned on the best solution.

adimoldovan replied:
OK, one more time.

> [!IMPORTANT]
> TLDR: There is nothing to fix here! The only was forward with this PR is to close it.




From @veljkho's comment, that video is from a failure in #53228. As I already mentioned this in this [comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1864239973) very early in this review, *that PR is still in progress*, and _**that test was expected to fail**_. 
In that run (and in that video) the flow was successful because the call to `/wp-json/wc/v3/products/` was actually successful. The test failed because in that commit the URL to be checked was wrong, it had an extra dot in front of it. Check the error: 
`response.url().includes( './wp-json/wc/v3/products/' )`. This was fixed in the next commit.

> [!IMPORTANT]
> Again, _**that is not a flakiness occurrence and should have been ignored in this investigation**_. All the other (few) failures of this test were caused by other parts of the code. This `waitForResponse` should be left alone.

---

This whole nonsense conversation could have been avoided if:
- a real investigation was actually performed, on all available failures, instead of simply removing some code based only on the last failure without even checking where it happened. Even it that response was the point of failure, removing it is not the fix!
- the feedback received very early would have beed actually read.
- there was no extra noise created by useless AI generated replies.

veljkho replied:
Please @adimoldovan , the video I was showing [was from this other PR and failure](https://github.com/woocommerce/woocommerce/actions/runs/12118305262/job/33782574373). The video is still there as a failure occurrence, which is the same as what I shared.

We are trying to solve flakiness in this test: `merchant/products/block-editor/create-grouped-product-block-editor.spec.js` as per [the issue ticket](https://github.com/woocommerce/woocommerce/issues/50820).

adimoldovan replied:
> Please @adimoldovan , the video I was showing [was from this other PR and failure](https://github.com/woocommerce/woocommerce/actions/runs/12118305262/job/33782574373). The video is still there as a failure occurrence, which is the same as what I shared.
> 
> We are trying to solve flakiness in this test: `merchant/products/block-editor/create-grouped-product-block-editor.spec.js` as per [the issue ticket](https://github.com/woocommerce/woocommerce/issues/50820).

Fine, my bad. Taking the step back, do we know why that response failed?

veljkho replied:
Please see my [observation](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867384010) from above.

It seems to me like a race condition issue where it attempted to wait for a response but was already received and failed to catch it. If we involve it earlier in the process, it will increase its chances of being caught. That's why I modified the code to support that. Please I need your expertise to confirm it, and I tested it literally 100 times locally. Also, for the UI-based assertion, if we could leave it as I really feel that is important for the E2E test (even if that maybe we could leave it without).

adimoldovan replied:
> If we involve it earlier in the process, it will increase its chances of being caught

Agreed. This looks to me like the real fix here.

> Also, for the UI-based assertion, if we could leave it as I really feel that is important for the E2E test (even if that maybe we could leave it without).

That assertion won't help with the fix, but it won't hurt the either. Maybe it creates another potential point of failure in the future, without it being a critical check? I can go either way on this, up to you.

veljkho replied:
> Agreed. This looks to me like the real fix here.

Happy to see this, thanks!

> Maybe it creates another potential point of failure in the future

My concern is that due to the very fast PW and the background process of saving a draft, verifying the notice check before proceeding further will make the test more stable and will add regression coverage that the saving effect should happen. This is especially good for external environments.

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js:
```diff
@@ -86,15 +86,19 @@ test.describe( 'General tab', { tag: '@gutenberg' }, () => {
-						response.status() === 200
-				);
 				await page.getByLabel( 'Categories' ).click();
-				await waitForCategoriesResponse;
+				await page
+					.getByLabel( 'Categories' )
+					.type( categoryName, { delay: 100 } );
+				await page
+					.getByRole( 'option', { name: categoryName } )
+					.locator( 'label' )
+					.waitFor();
+				await expect(
+					page
+						.getByRole( 'option', { name: categoryName } )
+						.locator( 'label' )
+				).toHaveText( categoryName );
```
There's probably some background processes in combination to a very fast Playwright that hides the category from showing up in the results at first point, so I added action to click and type the category name just like a normal end user would type, which adds some time and shows the category in the end. This way we also cover typing in the category field and searching for a category, which is nice regression to have.

adimoldovan replied:
I'm don't think removing `waitForCategoriesResponse` is a good idea. This is a very dynamic area and I remember having a lot of trouble stabilizing it and waitForCategoriesResponse fixed it. I see no harm in leaving it. And I'm very sure it's not a cause of flakiness, so why remove it?

adimoldovan replied:
> I added action to click and type the category name just like a normal end user would type, which adds some time and shows the category in the end.

Waiting for 100ms does nothing to guarantee the categories are loaded.
The `waitForCategoriesResponse` is there to make sure the categories are fetched before typing. If they are not, the test will fail.

veljkho replied:
I understand the reasoning behind keeping `waitForCategoriesResponse` — it worked in the past to stabilize the test, and I can see how that was necessary given the dynamic nature of this area. However, I encountered a specific issue where the test failed due to the timing of the response, and I believe waiting for the response alone wasn’t enough to guarantee the categories were loaded in time.

The test was failing intermittently while waiting for the response from waitForCategoriesResponse, and this was the root cause of the issue. The delay of 100ms between keystrokes I added to simulate typing was working well in stabilizing the input, but it wasn’t enough to handle the timing of the response in some cases.

As a result, I added a more explicit wait to ensure that the category option was not only typed into the field but also loaded and visible before interacting with it. The combination of waiting for the option’s label to appear and asserting the text seemed to be an effective way to prevent flaky failures, and I’ve executed it locally several times with consistent success.

I think the solution I implemented is more robust because it:

1. Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.
2. Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.

While `waitForCategoriesResponse` helped us in the past, I believe this additional wait for the category option’s label ensures better synchronization, reducing the chances of race conditions and preventing test failures.

I’m happy to discuss further if needed and hear from others as well.

adimoldovan replied:
> I understand the reasoning behind keeping `waitForCategoriesResponse` — it worked in the past to stabilize the test, and I can see how that was necessary given the dynamic nature of this area. However, I encountered a specific issue where the test failed due to the timing of the response, and I believe waiting for the response alone wasn’t enough to guarantee the categories were loaded in time.
> 
> The test was failing intermittently while waiting for the response from waitForCategoriesResponse, and this was the root cause of the issue. The delay of 100ms between keystrokes I added to simulate typing was working well in stabilizing the input, but it wasn’t enough to handle the timing of the response in some cases.
> 
> As a result, I added a more explicit wait to ensure that the category option was not only typed into the field but also loaded and visible before interacting with it. The combination of waiting for the option’s label to appear and asserting the text seemed to be an effective way to prevent flaky failures, and I’ve executed it locally several times with consistent success.
> 
> I think the solution I implemented is more robust because it:
> 
> 1. Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.
> 2. Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.
> 
> While `waitForCategoriesResponse` helped us in the past, I believe this additional wait for the category option’s label ensures better synchronization, reducing the chances of race conditions and preventing test failures.
> 
> I’m happy to discuss further if needed and hear from others as well.

Can you point me to at least one failure on this `waitForCategoriesResponse`? From what I could find there was a single failure (the last one) waiting for a response, but it was for the `/wp-json/wc/v3/products/`, at line 75, not this one. And the previous failures were for different causes. Please don't simply remove working code by guessing.

adimoldovan replied:
> * Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.

I'm sorry but this is not how it works. If you start typing before the categories are fetched it will fail. Typing will display the list of categories, and if the fetch finishes after the typing starts the list is not updated and the category will not be found.

> * Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.

This is not how waiting for the response works. There's nothing blind about it.

adimoldovan replied:
Uhm, this test `linked-product-tab-product-block-editor.spec.js` is not even reported as flaky. Why would you "fix" it? 
> The test was failing intermittently while waiting for the response from waitForCategoriesResponse

veljkho replied:
@adimoldovan , I have nothing against your implementation. I found it flaky and [here for example](https://github.com/woocommerce/woocommerce/actions/runs/12068850803/job/33655488067) it failed twice in a row. However, since we are dealing with such a low flakiness score and it is working well for the last many runs, let's just keep it as is. We could even combine both implementations if you'd like.

Please give it another look and let me know if I can merge it.

adimoldovan replied:
@veljkho 
1. That is a PR in progress. 
2. That is not a flaky occurrence. In that run the test failed - it is broken. Coincidently that is a PR I am working on and that failure was expected in that case.

Do you have a flakiness report on this test?

adimoldovan reviewed (changes requested):
@veljkho Please revisit my comments above. Not only these updates are not related to the flakiness, but they will make this test even more unstable.

veljkho commented:
Hi @adimoldovan 

Thanks for your feedback! I appreciate your perspective and understand your concerns about potentially introducing instability to the test. Let me clarify my reasoning further and address your points:

Not Related to Flakiness:

- The primary reason for my change was to address failures I observed related to timing, specifically when the categories were not fully loaded before the test attempted further actions. While `waitForCategoriesResponse` ensures the API request is completed, it doesn’t fully guarantee that the categories are rendered and interactable on the page.

- By explicitly waiting for the category option label to appear and verifying its content, I aimed to provide an additional layer of certainty that the UI is ready before interacting with it.

Concerns About Instability:

- I’d like to better understand how you feel these changes could make the test more unstable. From my testing, waiting for the DOM element directly has been more reliable, as it aligns the test flow with the actual state of the UI. If there are specific scenarios where this might cause instability, I’m open to discussing and finding a middle ground.

- My changes worked consistently in local tests and may add more resilience, especially in environments where UI rendering might lag behind the API response.

Proposed Next Steps:
- If we reencounter flakiness in the future, primarily related to timing or UI readiness, I suggest we combine both approaches. For now, let's use only `waitForCategoriesResponse`.

Ultimately, my goal is to ensure the test is as robust and stable as possible. I’m open to iterating on this further if needed.

veljkho commented:
@adimoldovan  I don't understand why are you behaving like this. Please review the code as I have added new commit. Thanks.

adimoldovan reviewed (changes requested):
> The flaky part was waiting for a response. I added a regression step instead of waiting for a response to fix the flakiness and increase coverage. 

I'm sorry but I don't agree. `waiting for a response` may be the point of failure, it doesn't mean it is the cause of flakiness. We need to first understand why the response failed. Removing the check for that response will not fix it.


Also see [my code comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1864256758)

veljkho commented:
Hi @adimoldovan 
I appreciate that you shared your thoughts, but I feel it’s important to address the tone of our conversation. While disagreement is a natural part of collaboration, how it’s communicated can make a big difference. I felt that your responses lacked respectful disagreement and constructive feedback, **which are essential for fostering a productive and collaborative environment**.

On the topic itself, I’d like to reiterate why `waitForResponse` should not replace UI-based assertions in E2E testing:

The primary purpose of e2e testing is to simulate real user interactions and validate that the application behaves as expected from the user’s perspective. This means relying on UI elements for assertions, as they directly reflect the user experience. Replacing `waitForResponse` with assertions based on UI elements ensures that the test reflects the actual experience users would have. This also means if further, the test fails by, e.g., adding products, we will be notified about it from an e2e perspective, and it could be a bug report (theoretically).

Overusing `waitForRespons` ties tests to specific API behaviors, making them fragile and prone to breaking with backend changes or delays, even when the user facing behavior is unaffected.

> Removing the check for that response will not fix it.
> I don't agree that this change will fix the flakiness. If the /wp-json/wc/v3/products/ request fails the notice will not be displayed either. We need to understand why the the /wp-json/wc/v3/products/failed first.

Here is the proof that the notice was indeed present when it failed to wait for a response:
- [**Please check this video**](https://d.pr/i/7uBZ3Z).

Failures in `waitForResponse` can be caused by backend issues, browser network conditions, or other external factors. These may not impact the user experience directly but can still cause the test to fail unnecessarily. By relying on UI checks, the test evaluates the final result that matters to the user, independent of backend instability.

Backend behavior is better validated through API tests. For example, if an API call fails, it should be addressed in a dedicated API test where its behavior can be isolated, rather than in an E2E test that focuses on the user interface.

I hope this summary clarifies my perspective and encourages a more collaborative approach moving forward. To address a prior remark, this long text is not an AI-generated response but **a well-structured and carefully considered perspective aimed at fostering constructive collaboration**.

veljkho commented:
Hey @adimoldovan 

Thank you for sharing your thoughts. I truly believe we’re both working toward the same goal: creating reliable and meaningful tests that benefit the team and the product. While our perspectives differ on this issue, I’ve made some adjustments to ensure the test is as robust and accurate as possible:

- I added a UI-based check to validate the presence of the notice and confirm that the user-facing behavior is correct.
- I modified the `waitForResponse` implementation to ensure the API response is received successfully, verifying that the backend process is functioning as expected.

By combining these two checks, I believe we can address both user experience and backend functionality concerns, making the test more comprehensive and reliable.

These updates have worked consistently for me locally, and I’m confident this approach aligns with best practices. However, your feedback is essential to ensure we’re on the right track. Could you please review the latest changes and let me know your thoughts?

Looking forward to collaborating further to finalize this test!

Commit:
Improve the e2e test to fix flakiness

juliaamosova reviewed (changes requested):
@veljkho I left a [comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867092458) related to your recent code changes. Please review and make further changes as discussed in the comment.

adimoldovan reviewed (approved):
Approving, as per [this comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867995887) the current code version looks like a good fix.
Thank you @veljkho.

juliaamosova reviewed (approved):
@veljkho thank you for your additional comments and reasoning behind your last change of moving the `waitForResponse` earlier in the test. I checked back API calls being made when the product editor to create a new product gets opened and there is indeed a call to get products. Catching that response earlier in the test makes sense to me as well. Approving this PR. I see that there are merge conflicts - please resolve the and we can then merge the PR.

juliaamosova commented:
>I see that there are merge conflicts - please resolve the and we can then merge the PR.
I don't see conflicts anymore. Merging.
